### ISSUE TITLE
About the descriptor loss

### ISSUE THREAD

**Issue by zhanggaojian** [eric-yyjau/pytorch-superpoint#10]
Hi,Thank you for your great work! What is the difference between the dense loss and the sparse loss?
I am looking forward to your reply.Thanks!

**Comment by eric-yyjau** [comment]
Hi, 
According to the paper "SuperPoint", we compute loss for positive pairs and negative pairs.
The differences lie in the number of samples.
For dense loss,
We sample all possible pairs. For positive pairs, we have around HxW pairs. For negative pairs, we have around HxW pairs for each pixel, which yields a total of (HxW)^2 pairs.
You can refer to https://github.com/eric-yyjau/pytorch-superpoint/blob/86b9621888e7c91801a5e9de889009573f407e45/utils/utils.py#L736 for implementation details.

For sparse loss (the default version here), we sample from the pool. We sample positive pairs: `num_matching_attempts: 600, negative pairs: `num_masked_non_matches_per_match: 100` to compute the contrastive loss. We use random sampling, no hard negative mining.

Let me know if you have further questions.

**Comment by zhanggaojian** [comment]
Thanks for your reply，it helps me a lot in understanding. I am continuing to learn your code.

**Comment by zhanggaojian** [comment]
@eric-yyjau   Hi, i am sorry to bother you again. The single GPU is used to train the superpoint on CoCo. When I run the training script, the error emerges as follow, so I beg to your help.

`python train4.py train_joint configs/superpoint_coco_train_heatmap.yaml superpoint_coco --eval `--debug`

` File "train4.py", line 93, in train_joint
    train_agent.train()
  File "/home/phantom/Downloads/zgj/pytorch-superpoint-master/Train_model_frontend.py", line 277, in train
    loss_out = self.train_val_sample(sample_train, self.n_iter, True)
  File "/home/phantom/Downloads/zgj/pytorch-superpoint-master/Train_model_heatmap.py", line 315, in train_val_sample
    **self.desc_params
  File "/home/phantom/Downloads/zgj/pytorch-superpoint-master/utils/loss_functions/sparse_loss.py", line 243, in batch_descriptor_loss_sparse
    homographies[i].type(torch.float32), **options)
  File "/home/phantom/Downloads/zgj/pytorch-superpoint-master/utils/loss_functions/sparse_loss.py", line 212, in descriptor_loss_sparse
    num_masked_non_matches_per_match=num_masked_non_matches_per_match)
  File "/home/phantom/Downloads/zgj/pytorch-superpoint-master/utils/loss_functions/sparse_loss.py", line 107, in get_non_matches_corr
    img_b_mask=None)
  File "/home/phantom/Downloads/zgj/pytorch-superpoint-master/utils/correspondence_tools/correspondence_finder.py", line 261, in create_non_correspondences
    diffs_0_flattened = diffs_0.view(-1,1)
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
`

**Comment by eric-yyjau** [comment]
Hi, I'm sorry that you're bothered by the issues.
Some guesses.
1. Do you mind check the dataloader is working correctly? The shape and type of the input/ output data.
2. Can you check the gpu is assigned correctly?
3. Did it happen right after you started training?
Thanks.

**Comment by zhanggaojian** [comment]
Thank you very much. The problem was solved by your reference.

- [ ] 1

**Comment by zhanggaojian** [comment]
Thank you very much. The problem was solved by the reference you provided.

I change code: utils/correspondence_tools/correspondence_finder.py , line 261 and 262, add contiguous() before view()

**Comment by eric-yyjau** [comment]
Sounds good! Thank you for your feedback.

**Comment by ddetone** [comment]
In the original implementation, we actually used a "semi-dense" loss, where the loss is computed at the center of each 8x8 cell center, resulting in (H/8)*(W/8) positive pairs and ((H/8)*(W/8))^2 negative pairs. This creates a nice bias that helps the descriptor reject false matches, however, it requires a lot of memory, and becomes restrictive for resolutions beyond 320x240.

To train at higher resolutions (640x480 and beyond), I found using the sparse loss as a better option.
